File: CPP VW/CPP VW/Animal.cpp
#include <iostream>
#include "Animal.h"
#include <time.h>

Animal::Animal(World& worldref, int x, int y) :
	Organism(worldref, x, y)
{
}
void Animal::Action() {
	if (this->specialAction())
		return;
	this->Move();
};
void Animal::Collision(Organism& Organism) {};
Animal::~Animal() {};

bool Animal::Multiply(int posX, int posY) {
	for (int y = -1; y < 2; y++) {
		for (int x = -1; x < 2; x++) {
			if (world.InBorders(posX + x, posY + y) && world.IsEmpty(posX + x, posY + y)) {
				world.AddOrganism(posX + x, posY + y, this->GetSign());
				return true;
			}
		}
	}
	return false;
};
void Animal::Move(int fields) {
	int x, y;
	do {
		x = (rand() % 3 - 1)*fields;
		y = (rand() % 3 - 1)*fields;
	} while (!world.InBorders(this->posX + x, this->posY + y) || x==0 && y==0);
		world.WorldMove(this->posX, this->posY, this->posX + x, this->posY + y);
};

File: JAVA VW/src/programowanie/obiektowe/proj/Animal.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package programowanie.obiektowe.proj;

/**
 *
 * @author kubaz
 */
 public class Animal extends Organism{

    public Animal(Position position, World world) {
        super(position, world);
    }
    
    @Override
    protected void multiply(){
        Position newLoc = world.freeRandomAround(location, 1);
        if(newLoc.isEqual(this.location))return;
        world.addOrganism(newLoc, this.getName());
        System.out.println(this.getName()+ " multiplies");
    }
    
    @Override
    public void action(){
       this.movePermission=true;
       move(world.randomAround(this.location, 1));
    }
    
     protected void move (Position newPos){
        if(newPos.isEqual(location))return;
        if(!world.inBoundaries(newPos))return;
        if(world.isEmpty(newPos)){
            location=newPos;
            return;
        }
        Organism enemy = world.findObject(newPos);
        
        if(!enemy.isKilled())
            collision(this, enemy);
        
        if(!this.killed && this.movePermission){
            location=newPos;
        }
    }
}
    